extensions/resource-deployment/src/services/platformService.ts

```diff
@@ -76,19 +97,106 @@ export class PlatformService implements IPlatformService {
 		return fs.promises.readFile(filePath, 'utf8');
 	}
 
-	runCommand(command: string, options?: CommandOptions): Promise<string> {
+	public logToOutputChannel(data: string | Buffer, header?: string): void {
+		data.toString().split(/\r?\n/)
+			.forEach(line => {
+				this._outputChannel.appendLine(header ? header + line : line);
+			});
+	}
+
+	private outputDataChunk(data: string | Buffer, outputChannel: vscode.OutputChannel, header: string): void {
+		data.toString().split(/\r?\n/)
+			.forEach(line => {
+				outputChannel.appendLine(header + line);
+			});
+	}
+
+	async runCommand(command: string, options?: CommandOptions, sudo?: boolean, commandTitle?: string): Promise<string> {
+		if (commandTitle !== undefined && commandTitle !== null) {
+			this._outputChannel.appendLine(`\t<<<${commandTitle}>>>`);
+		}
+		console.log(`TCL: PlatformService -> options: ${JSON.stringify(options)}`);
+		if (sudo) {
+			return this.runSudoCommand(command, options, this._outputChannel);
+		} else {
+			return this.runStreamedCommand(command, options, this._outputChannel);
+		}
+	}
+
+	private runSudoCommand(cmd: string, options: CommandOptions | undefined, outputChannel: vscode.OutputChannel): Promise<string> {
 		return new Promise<string>((resolve, reject) => {
-			const env = Object.assign({}, process.env, options && options.additionalEnvironmentVariables);
-			cp.exec(command, {
-				cwd: options && options.workingDirectory,
+			if (outputChannel) {
+				outputChannel.appendLine(`    > ${cmd}`);
+			}
+
+			if (options && options.workingDirectory) {
+				process.chdir(options.workingDirectory);
+			}
+
+			// Workaround for https://github.com/jorangreef/sudo-prompt/issues/111
+			const origEnv: NodeJS.ProcessEnv = Object.assign({}, process.env, options && options.additionalEnvironmentVariables);
+			const env: NodeJS.ProcessEnv = {};
+
+			Object.keys(origEnv).filter(key => /^[a-zA-Z_][a-zA-Z0-9_]*$/.test(key)).forEach((key) => {
+				env[key] = origEnv[key];
+			});
+
+			const sudoOptions = {
+				name: sudoPromptTitle,
 				env: env
-			}, (error, stdout, stderror) => {
+			};
+			sudo.exec(cmd, sudoOptions, (error, stdout, stderr) => {
 				if (error) {
+					this.outputDataChunk(error, outputChannel, '    stderr: ');
 					reject(error);
 				} else {
+					if (stdout) {
+						this.outputDataChunk(stdout, outputChannel, '    stdout: ');
+					}
+					if (stderr) {
+						this.outputDataChunk(stderr, outputChannel, '    stderr: ');
+					}
 					resolve(stdout);
 				}
 			});
 		});
 	}
+
+	private runStreamedCommand(cmd: string, options?: CommandOptions, outputChannel?: vscode.OutputChannel): Promise<string> {
+		return new Promise<string>((resolve, reject) => {
```

Comment:
Do you have to return them like this? Usually promises written like this are a source of bugs in future. For example a lot of the OE bugs were solved by rewriting to not use this syntax.

Reply:
For context, when going from functions that expect callbacks to the Promise-ified world this is fairly standard practice. I've had to do this elsewhere when dealing with streams and I don't know of a better way. There is definitely higher risk so if you know of cleaner promise-friendly ways to work with spawn / child processes I'm happy to get that adopted.

Reply:
It's okay to use custom promises, especially in situations when dealing with streams and such. 

Also, as discussed in DRI these reviews aren't meant to be something like "I'm right and you should do it this way", its more of "hey there is a potential pitfall here"

As long as the promise constructor isn't being misused, it's okay to use it.